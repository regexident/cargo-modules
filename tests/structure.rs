#[macro_use]
mod util;

mod help {
    test_cmd!(
        args: "structure \
                --help",
        success: true,
        color_mode: ColorMode::Plain,
        project: smoke_structure
    );
}

mod colors {
    mod plain {
        test_cmd!(
            args: "structure",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke_structure
        );
    }
}

mod default {
    mod pass {
        test_cmds!(
            args: "structure",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                package_lib_target,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_bin_target,
                workspace_single_package_lib_target,
                workspace_single_package_bin_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "structure",
            success: false,
            color_mode: ColorMode::Plain,
            projects: [
                package_multi_target,
                virtual_workspace_single_package_multi_target,
                workspace_single_package_multi_target,
            ]
        );
    }
}

mod lib {
    mod pass {
        test_cmds!(
            args: "structure \
                    --lib",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_lib_target,
                package_multi_target,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_multi_target,
                workspace_single_package_lib_target,
                workspace_single_package_multi_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "structure \
                    --lib", // does not exist
            success: false,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                virtual_workspace_single_package_bin_target,
                workspace_single_package_bin_target,
            ]
        );
    }
}

mod bin {
    mod pass {
        test_cmds!(
            args: "structure \
                    --bin package_bin_target",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                virtual_workspace_single_package_bin_target,
                workspace_single_package_bin_target,
            ]
        );

        test_cmds!(
            args: "structure \
                    --bin package_multi_target",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_multi_target,
                virtual_workspace_single_package_multi_target,
                workspace_single_package_multi_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "structure \
                    --bin foobar", // does not exist
            success: false,
            color_mode: ColorMode::Plain,
            projects: [
                package_lib_target,
                package_multi_target,
                virtual_workspace_multi_package,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_multi_target,
                workspace_multi_package,
                workspace_single_package_lib_target,
                workspace_single_package_multi_target,
            ]
        );
    }
}

mod package {
    mod pass {
        test_cmds!(
            args: "structure \
                    --package package_lib_target",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_lib_target,
                virtual_workspace_single_package_lib_target,
                workspace_single_package_lib_target,
            ]
        );

        test_cmds!(
            args: "structure \
                    --package package_bin_target",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                virtual_workspace_single_package_bin_target,
                workspace_single_package_bin_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "structure \
                    --package foobar",
            success: false,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                package_lib_target,
                package_multi_target,
                virtual_workspace_multi_package,
                virtual_workspace_single_package_bin_target,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_multi_target,
                workspace_multi_package,
                workspace_single_package_bin_target,
                workspace_single_package_lib_target,
                workspace_single_package_multi_target,
            ]
        );
    }
}

mod package_lib {
    mod pass {
        test_cmds!(
            args: "structure \
                    --package package_lib_target \
                    --lib",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_lib_target,
                workspace_single_package_lib_target,
                virtual_workspace_single_package_lib_target,
            ]
        );

        test_cmds!(
            args: "structure \
                    --package package_multi_target \
                    --lib",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_multi_target,
                workspace_single_package_multi_target,
                virtual_workspace_single_package_multi_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "structure \
                    --package package_bin_target \
                    --lib", // does not exist
            success: false,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                workspace_single_package_bin_target,
                virtual_workspace_single_package_bin_target,
            ]
        );
    }
}

mod package_bin {
    mod pass {
        test_cmds!(
            args: "structure \
                    --package package_bin_target \
                    --bin package_bin_target",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                workspace_single_package_bin_target,
                virtual_workspace_single_package_bin_target,
            ]
        );

        test_cmds!(
            args: "structure \
                    --package package_multi_target \
                    --bin package_multi_target",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_multi_target,
                workspace_single_package_multi_target,
                virtual_workspace_single_package_multi_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "structure \
                    --package workspace-package \
                    --bin foobar", // does not exist
            success: false,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                package_lib_target,
                package_multi_target,
                virtual_workspace_multi_package,
                virtual_workspace_single_package_bin_target,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_multi_target,
                workspace_multi_package,
                workspace_single_package_bin_target,
                workspace_single_package_lib_target,
                workspace_single_package_multi_target,
            ]
        );
    }
}

mod cfg_test {
    test_cmd!(
        args: "structure \
                --cfg-test",
        success: true,
        color_mode: ColorMode::Plain,
        project: smoke_structure
    );

    mod without_tests {
        test_cmd!(
            args: "structure",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke_structure
        );
    }

    mod with_tests {
        test_cmd!(
            args: "structure \
                    --cfg-test",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke_structure
        );
    }
}

mod selection {
    mod no_fns {
        test_cmd!(
            args: "structure \
                    --no-fns",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke_structure
        );
    }

    mod no_traits {
        test_cmd!(
            args: "structure \
                    --no-traits",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke_structure
        );
    }

    mod no_types {
        test_cmd!(
            args: "structure \
                    --no-types",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke_structure
        );
    }
}

mod focus_on {
    mod simple_path {
        test_cmd!(
            args: "structure \
                    --focus-on \"crate::lorem::sit\"",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke_structure
        );
    }

    mod glob_path {
        test_cmd!(
            args: "structure \
                    --focus-on \"crate::lorem::sit::*\"",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke_structure
        );
    }

    mod self_path {
        test_cmd!(
            args: "structure \
                    --focus-on \"crate::lorem::sit::{self}\"",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke_structure
        );
    }

    mod use_tree {
        test_cmd!(
            args: "structure \
                    --focus-on \"crate::{lorem::sit::ipsum, sit::dolor::amet::lorem::sit}\"",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke_structure
        );
    }

    mod with_max_depth {
        test_cmd!(
            args: "structure \
                    --focus-on \"crate::{lorem, sit}\" \
                    --max-depth 2",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke_structure
        );
    }

    mod nonexistent_path {
        test_cmd!(
            args: "structure \
                    --focus-on \"nonexistent\"",
            success: false,
            color_mode: ColorMode::Plain,
            project: smoke_structure
        );
    }
}

mod max_depth {
    mod depth_0 {
        test_cmd!(
            args: "structure \
                    --no-types \
                    --no-traits \
                    --no-fns \
                    --max-depth 0",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke_structure
        );
    }

    mod depth_1 {
        test_cmd!(
            args: "structure \
                    --no-types \
                    --no-traits \
                    --no-fns \
                    --max-depth 1",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke_structure
        );
    }

    mod depth_2 {
        test_cmd!(
            args: "structure \
                    --no-types \
                    --no-traits \
                    --no-fns \
                    --max-depth 2",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke_structure
        );
    }
}

mod sort_by {
    mod name {
        test_cmd!(
            args: "structure \
                    --sort-by \"name\"",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke_structure
        );
    }

    mod visibility {
        test_cmd!(
            args: "structure \
                    --sort-by \"visibility\"",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke_structure
        );
    }

    mod kind {
        test_cmd!(
            args: "structure \
                    --sort-by \"kind\"",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke_structure
        );
    }
}

mod sort_reversed {
    test_cmd!(
        args: "structure \
                --sort-reversed",
        success: true,
        color_mode: ColorMode::Plain,
        project: smoke_structure
    );
}

mod github {
    mod issue_80 {
        mod tests {
            test_cmd!(
                args: "structure \
                        --cfg-test \
                        --no-traits \
                        --no-fns",
                success: true,
                color_mode: ColorMode::Plain,
                project: github_issue_80
            );
        }

        mod without_tests {
            test_cmd!(
                args: "structure \
                        --no-traits \
                        --no-fns",
                success: true,
                color_mode: ColorMode::Plain,
                project: github_issue_80
            );
        }
    }

    mod issue_222 {
        test_cmd!(
            args: "structure \
                    --no-types \
                    --no-traits \
                    --no-fns",
            success: true,
            color_mode: ColorMode::Plain,
            project: github_issue_222
        );
    }

    mod issue_362 {
        test_cmd!(
            args: "structure \
                    --features 'opt-in'",
            success: true,
            color_mode: ColorMode::Plain,
            project: github_issue_362
        );
    }
}
