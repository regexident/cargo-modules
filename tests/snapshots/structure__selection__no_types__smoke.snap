---
source: tests/structure.rs
expression: output
---
CMD:
structure
--no-types

STDERR:

STDOUT:

crate smoke
├── mod derives: pub(crate)
├── mod functions: pub(crate)
│   ├── fn body: pub(self)
│   ├── fn inputs: pub(self)
│   └── fn outputs: pub(self)
├── mod hierarchy: pub(crate)
│   └── mod lorem: pub(self)
│       ├── mod consectetur: pub(self)
│       │   └── mod adipiscing: pub(self)
│       │       └── mod elit: pub(self)
│       ├── mod dolor: pub(self)
│       │   └── mod sit: pub(self)
│       │       └── mod amet: pub(self)
│       └── mod ipsum: pub(self)
├── mod methods: pub(crate)
├── mod target: pub(crate)
├── mod uses: pub(crate)
│   └── mod cycle: pub(self)
│       ├── mod node_0: pub(self)
│       └── mod node_1: pub(self)
│           └── mod node_2: pub(self)
└── mod visibility: pub(crate)
    └── mod dummy: pub(self)
        ├── mod enums: pub(self)
        ├── mod fns: pub(self)
        │   ├── fn pub_crate: pub(crate)
        │   ├── fn pub_module: pub(in crate::visibility)
        │   ├── fn pub_private: pub(self)
        │   ├── fn pub_public: pub
        │   └── fn pub_super: pub(super)
        ├── mod kinds: pub(self)
        │   ├── async fn AsyncFunction: pub(self)
        │   ├── const fn ConstFunction: pub(self)
        │   ├── fn Function: pub(self)
        │   ├── mod Module: pub(self)
        │   ├── trait Trait: pub(self)
        │   ├── unsafe fn UnsafeFunction: pub(self)
        │   └── unsafe trait UnsafeTrait: pub(self)
        ├── mod mods: pub(self)
        │   ├── mod pub_crate: pub(crate)
        │   ├── mod pub_module: pub(in crate::visibility)
        │   ├── mod pub_private: pub(self)
        │   ├── mod pub_public: pub
        │   └── mod pub_super: pub(super)
        ├── mod structs: pub(self)
        ├── mod traits: pub(self)
        │   ├── mod r#unsafe: pub(self)
        │   │   ├── unsafe trait PubCrate: pub(crate)
        │   │   ├── unsafe trait PubModule: pub(in crate::visibility)
        │   │   ├── unsafe trait PubPrivate: pub(self)
        │   │   ├── unsafe trait PubPublic: pub
        │   │   └── unsafe trait PubSuper: pub(super)
        │   └── mod safe: pub(self)
        │       ├── trait PubCrate: pub(crate)
        │       ├── trait PubModule: pub(in crate::visibility)
        │       ├── trait PubPrivate: pub(self)
        │       ├── trait PubPublic: pub
        │       └── trait PubSuper: pub(super)
        └── mod unions: pub(self)
